sql/expression/function/ceil_round_floor.go

```diff
@@ -0,0 +1,280 @@
+	if sql.IsNumber(childType) {
+		return childType
+	}
+	return sql.Int32
+}
+
+func (c *Ceil) String() string {
+	return fmt.Sprintf("CEIL(%s)", c.Child)
+}
+
+// TransformUp implements the Expression interface.
+func (c *Ceil) TransformUp(fn sql.TransformExprFunc) (sql.Expression, error) {
+	child, err := c.Child.TransformUp(fn)
+	if err != nil {
+		return nil, err
+	}
+	return fn(NewCeil(child))
+}
+
+// Eval implements the Expression interface.
+func (c *Ceil) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	child, err := c.Child.Eval(ctx, row)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if child == nil {
+		return nil, nil
+	}
+
+	if !sql.IsNumber(c.Child.Type()) {
+		return 0, nil
+	}
+
+	if !sql.IsDecimal(c.Child.Type()) {
+		return child, err
+	}
+
+	switch num := child.(type) {
+	case float64:
+		return math.Ceil(num), nil
+	case float32:
+		return float32(math.Ceil(float64(num))), nil
+	default:
+		return nil, sql.ErrInvalidType.New(reflect.TypeOf(num))
+	}
+}
+
+// Floor returns the biggest integer value not less than X.
+type Floor struct {
+	expression.UnaryExpression
+}
+
+// NewFloor returns a new Floor expression.
+func NewFloor(num sql.Expression) sql.Expression {
+	return &Floor{expression.UnaryExpression{Child: num}}
+}
+
+// Type implements the Expression interface.
+func (f *Floor) Type() sql.Type {
+	childType := f.Child.Type()
+	if sql.IsNumber(childType) {
+		return childType
+	}
+	return sql.Int32
+}
+
+func (f *Floor) String() string {
+	return fmt.Sprintf("FLOOR(%s)", f.Child)
+}
+
+// TransformUp implements the Expression interface.
+func (f *Floor) TransformUp(fn sql.TransformExprFunc) (sql.Expression, error) {
+	child, err := f.Child.TransformUp(fn)
+	if err != nil {
+		return nil, err
+	}
+	return fn(NewFloor(child))
+}
+
+// Eval implements the Expression interface.
+func (f *Floor) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	child, err := f.Child.Eval(ctx, row)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if child == nil {
+		return nil, nil
+	}
+
+	if !sql.IsNumber(f.Child.Type()) {
+		return 0, nil
+	}
+
+	if !sql.IsDecimal(f.Child.Type()) {
+		return child, err
+	}
+
+	switch num := child.(type) {
+	case float64:
+		return math.Floor(num), nil
+	case float32:
+		return float32(math.Floor(float64(num))), nil
+	default:
+		return nil, sql.ErrInvalidType.New(reflect.TypeOf(num))
+	}
+}
+
+// Round returns the number (x) with (d) requested decimal places.
+// If d is negative, the number is returned with the (abs(d)) least significant
+// digits of it's integer part set to 0. If d is not specified or nil/null
+// it defaults to 0.
+type Round struct {
+	expression.BinaryExpression
+}
+
+// NewRound returns a new Round expression.
+func NewRound(args ...sql.Expression) (sql.Expression, error) {
+	argLen := len(args)
+	if argLen == 0 || argLen > 2 {
+		return nil, sql.ErrInvalidArgumentNumber.New("1 or 2", argLen)
+	}
+
+	d := func(i int) sql.Expression {
```

Comment:
What do you think about this for avoiding that extra closure?

```go
var right sql.Expression = nil
if len(args) == 2 {
        right = args[1]
}
```

Reply:
In favor of performance, yes.

(I am too influenced by other languages nowdays..)